Add support for the encoding flag.

vds2212 opened the pull request:
The `rg` runtime support the `-E` flag that let the user specify the expected encoding of the file to be searched. The default is `utf-8` but if the file encoding is another one (e.g.: `latin1`) the flag let inform `rg` that the file should be considered encoded as `latin1` this is particularly useful to:
- search accentuated strings
- display in the result buffer the accentuated strings

vds2212 committed:
support -E flag determine the encoding

vds2212 committed:
Add documentation for the -encoding flag.

vds2212 committed:
ensure a default value g:ctrlsf_encoding is provided

dyng commented on autoload/ctrlsf/win.vim:
```diff
@@ -145,6 +145,8 @@ endf
 "
 func! ctrlsf#win#SetModifiableByViewMode(modifiable) abort
+    call setbufvar(s:MAIN_BUF_NAME, '&modifiable', 1)
+    call setbufvar(s:MAIN_BUF_NAME, '&fileencoding', g:ctrlsf_encoding)
```
To make things simpler, I prefer keeping ctrlsf buffer's encoding to `utf-8`. You could iconv characters in other encoding into utf8 during rendering.

vds2212 commented on autoload/ctrlsf/opt.vim:
```diff
@@ -291,7 +294,7 @@ func! s:ParseOptions(options_str) abort
             let options[name] = 1
         elseif opt.args == 1
-            if tokens[i] =~# '\d\+'
+            if tokens[i] =~# '^\d\+'
```
The reason I change that line is that `\d\+` match `utf8` and then the code was converting the option to `0` (`str2nr('utf8') == 0`) which was making the code to fail.
The change propose to only convert to integer the options that are integer.
A better proposition could even be `^\d\+$`.

vds2212 committed:
Result Review:

Integrate the result of the review of dyng:
- Get rid of the g:ctrlsf_encoding variable
- Use of the ctrlsf#opt#GetOpt('encoding')

vds2212 commented:
Thanks for the nice review.

I have get rid of the `g:ctrlsf_encoding` variable and use the `ctrlsf#opt#GetOpt('encoding)` as suggested.

The only suggested that I didn't took is the `\d+` since for `utf-8` or `latin1` it convert wrongly the option into a number.

dyng commented on autoload/ctrlsf/edit.vim:
```diff
@@ -89,7 +89,9 @@ func! s:VerifyConsistent(on_disk, on_mem) abort
 
-            if line.content !=# a:on_disk[ln-1]
+            let content = line.content
+            let content = iconv(content, 'utf-8', ctrlsf#opt#GetOpt('encoding'))
```
I think it's better to perform iconv only if encoding is provided.

dyng commented on autoload/ctrlsf/edit.vim:
```diff
@@ -114,10 +116,12 @@ func! s:WriteParagraph(buffer, orig, modi) abort
         let ln    = start_lnum + i
 
+        let line = mline.content
+        let line = iconv(line, 'utf-8', ctrlsf#opt#GetOpt('encoding'))
```
Save as above.

vds2212 commented on autoload/ctrlsf/win.vim:
```diff
@@ -145,6 +145,7 @@ endf
 " SetModifiable()
 "
 func! ctrlsf#win#SetModifiableByViewMode(modifiable) abort
+    call setbufvar(s:MAIN_BUF_NAME, '&modifiable', 1)
```
I was necessary to change the encoding of the buffer. Since we don't change it anymore. I must remove it too 👍

vds2212 committed:
Second review session:

- Use encoding only if specified
- Convert option to integer if option match ^\d\+$ instead of \d\+

vds2212 commented:
I committed a new version that incorporate all the changes you suggested :-)
Thanks again for your review!

dyng commented:
@vds2212 merged. Thank you for your contribution!

vds2212 commented:
Thanks to you for your review and of course for this very nice plugin!
